File: puzzle-01/solution.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func fileToLines(addr string) []string {
	content, err := os.ReadFile(addr)
	if err != nil {
		return nil
	}

	lines := strings.Split(string(content), "\n")

	var result []string
	for _, line := range lines {
		if line != "" {
			result = append(result, line)
		}
	}
	return lines
}

var wordMap = map[string]int{
	"one":   1,
	"two":   2,
	"three": 3,
	"four":  4,
	"five":  5,
	"six":   6,
	"seven": 7,
	"eight": 8,
	"nine":  9,
	"zero":  0,
	"1":     1,
	"2":     2,
	"3":     3,
	"4":     4,
	"5":     5,
	"6":     6,
	"7":     7,
	"8":     8,
	"9":     9,
	"0":     0,
}

type Chunk struct {
	five  string
	four  string
	three string
	digit string
}

func slidingWindow(line string) int {
	numbers := make([]int, 0)
	line = "   " + line + "     "
	length := len(line)
	for i := 0; i <= length-5; i++ {
		chunk := Chunk{
			five:  line[i : i+5],
			four:  line[i : i+4],
			three: line[i : i+3],
			digit: line[i : i+1],
		}
		if res, ok := wordMap[chunk.five]; ok {
			numbers = append(numbers, res)
		}
		if res, ok := wordMap[chunk.four]; ok {
			numbers = append(numbers, res)
		}
		if res, ok := wordMap[chunk.three]; ok {
			numbers = append(numbers, res)
		}
		if res, ok := wordMap[chunk.digit]; ok {
			numbers = append(numbers, res)
		}
	}
	first := numbers[0]
	last := numbers[len(numbers)-1]
	return first*10 + last
}


func main() {
	lines := fileToLines("input.txt")
	numbers := make([]int, 0, len(lines))
	for _, line := range lines {
		n := slidingWindow(line)
		numbers = append(numbers, n)
	}

	sum := int(0)
	for _, n := range numbers {
		sum += n
	}
	fmt.Println(sum)
}

File: puzzle-01/solution.py
with open('input2.txt','r') as f:
    lines = f.readlines()
lines = [l[:len(l)-1] for l in lines]
print(lines[-10:])

def check_for_words(line,i,reverse=False):
    if reverse:
        c3 = line[i-2:i+1]
        c4 = line[i-3:i+1]
        c5 = line[i-4:i+1]
    else:
        c3 = line[i:i+3]
        c4 = line[i:i+4]
        c5 = line[i:i+5]
    print(c3,c4,c5)
    if c3 == 'one':
        return 1
    if c3 == 'two':
        return 2
    if c3=='six':
        return 6
    if c4=='four':
        return 4
    if c4=='five':
        return 5
    if c4=='nine':
        return 9
    if c5=='three':
        return 3
    if c5=='seven':
        return 7
    if c5=='eight':
        return 8
    return None
    


values = []

for l in lines:
    a,b = 0,0
    for i,char in enumerate(l):
        if char in ['0','1','2','3','4','5','6','7','8','9']:
            a = int(char)
            break
        if char in ['o','t','f','s','e','n']:
            res = check_for_words(l,i)
            if res is not None:
                a = res
                break
    
    for i,char in enumerate(l[::-1]):
        if char in ['0','1','2','3','4','5','6','7','8','9']:
            b = int(char)
            break

        if char in ['e','o','r','x','n','t']:
            res = check_for_words(l,i,reverse=True)
            if res is not None:
                b = res
                break
    values.append(a*10+b)

print(values)

print(sum(values))